File: ruby/bob/bob.rb
# The lackadaisical teenager
class Bob
  class << self
    def hey(remark)
      responses.keys.each do |type|
        return responses[type] if send("#{type}?", remark.strip)
      end
    end

    private

    def responses
      {
        empty: 'Fine. Be that way!',
        question_yell: 'Calm down, I know what I\'m doing!',
        question: 'Sure.',
        yell: 'Whoa, chill out!',
        others: 'Whatever.'
      }
    end

    def question?(remark)
      remark.end_with?('?')
    end

    def yell?(remark)
      (remark.upcase == remark) && (remark.count('A-Z') > 0)
    end

    def question_yell?(remark)
      question?(remark) && yell?(remark)
    end

    def empty?(remark)
      remark.empty?
    end

    def others?(_remark)
      true
    end
  end
end

p Bob.hey("")

File: elixir/bob/lib/bob.ex
defmodule Bob do
   def hey(input) do
    input = String.trim input
    cond do
      nothing? input -> "Fine. Be that way!"
      yell_question? input -> "Calm down, I know what I'm doing!"
      question? input -> "Sure."
      yell? input -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end

  defp question? input do
    String.ends_with? input, "?"
  end

  defp nothing? input do
    String.length(input) == 0
  end

  defp yell? input do
    String.upcase(input) == input && String.match?(input, ~r/\p{L}/)
  end

  defp yell_question? input do
    yell?(input) && question?(input)
  end
end
